File: Python/burbuja_centinela.py
# Programa para buscar un elemento mediante centinela en una lista y ordenarla usando ordenamiento por intercambio burbujas

# Funcion para rellenar una lista

def fill_list(lista):
    length = int(input("\nIngrese la longitud de la lista: "))

    for i in range(0, length):
        element = int(input(f"\nIngresa el valor del elemento {i}: "))
        lista.append(element)

# Funcion para la busqueda por centinela

def sentinel_search(lista, size, v):
    lista.append(v)
    i = 0

    while (lista[i] != v):
        i += 1;

    if i < size:
        return i
    else:
        return -1

# Funcion para el bubble sort

def bubble_sort(lista, size):
    print("\nOrdenando por bubble sort")

    for i in range(0, size - 1):
        for j in range(0, size - i - 1):
            if lista[j] > lista[j + 1]:
                aux = lista[j]
                lista[j] = lista[j + 1]
                lista[j + 1] = aux



# MAIN PROGRAM

int_list = [];

# Llamada a funcion para rellenar la lista

fill_list(int_list)
print(f"\nLa lista es: {int_list}")

# Llamada a funcion para la busqueda por centinela

length = len(int_list)
value = int(input("\nIngrese el valor del elemento a buscar: "))
pos = sentinel_search(int_list, length, value)
print(f"\nLa posicion del elemento {value} es {pos}")

# Eliminando el centinela de la lista

int_list.pop()

# Llamada a funcion para el bubble sort

bubble_sort(int_list, length)
print(f"\nLa lista ordenada es: {int_list}")

File: C/burbuja_centinela.c
// Programa para implementar la busqueda por centinela y ordenar el arreglo usando el algoritmo del bubble sort

#include <stdio.h>

// Funcion para llenar un arreglo

void fill_array(int arr[], int size)
{
    register int i;
    int element;

    for (i = 0; i < size; i++)
    {
        printf("\nIngrese el valor del elemento %d ", i);
        scanf("%d", &element);
        arr[i] = element;
    }
    printf("\n");
}

// Funcion para imprimir un arreglo

void print_array(int arr[], int size)
{
    register int i;

    printf("\nLos elementos del arreglo son: [ ");
    
    for (i = 0; i < size; i++)
    {
        printf("%d ",arr[i]);
    }
    
    printf("]\n");
}

// Funcion para la busqueda por centinela

int sentinel_search(int arr[], int size, int v)
{
    register int i;

    for (i = 0, arr[size] = v; arr[i] != v; i++);

    if (i < size)
        return i;
    else
        return -1;
}

// Funcion para el bubble sort

void bubble_sort(int arr[], int size)
{
    register int i, j;

    for (i = 0; i < size - 1; i++)
    {
        for (j = 0; j < size - i - 1; j++)
        {
            if (arr[j] > arr[j + 1])
            {
                int aux;
                aux = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = aux;
            }
        }
    }
}

void main()
{
    int length, pos, value, n;
    
    printf("\nIngrese la cantidad de elementos del arreglo: ");
    scanf("%d", &n);

    int array[n];

    length = sizeof(array)/sizeof(array[0]);

    // Llamada a funcion para rellenar el arreglo

    fill_array(array, length);
    print_array(array, length);

    // Llamada a funcion para la busqueda por centinela

    printf("\nIngrese el elemento que desea buscar: ");
    scanf("%d", &value);

    pos = sentinel_search(array, length, value);
    printf("\nLa posicion del elemento %d es %d\n", value, pos);

    // Llamada a funcion para ordenar por bubble sort

    bubble_sort(array, length);
    print_array(array, length);
}